xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h

```diff
@@ -416,6 +461,15 @@ struct AttentionBackwardKernel {
     using Mma =
         typename MakeCustomMma<typename DefaultMma::ThreadblockMma, kMaxK>::Mma;
 
+    // used for efficient load of bias tile (Bij) from global memory to shared memory
+    using BiasLoader = TileSmemLoader<
+        scalar_t,
+        // applied to transposed attn matrix tile (Pij), so we load Bij transposed too
+        cutlass::MatrixShape<ThreadblockShape::kN, ThreadblockShape::kM>,
+        MmaCore::kThreads,
+        // input restriction: kv_len has to be a multiple of this value
```

Comment:
> First, that's a pretty impressive PR and significant changes, which must have taken a lot of effort :o
Having attention bias (and in a smaller measure dropout) is something we've wanted to implement for some time but didn't have time to work on, so this contribution is very welcome!
I'll try to take a look at this in the next few days - as it's a pretty big chunk.

thanks! sorry about the large PR, i sort of underestimated the changes required and did everything at once

> (1) Is there any performance regression for the users who don't use a mask/dropout (mostly worried for V100/A100 - I'll run benchmarks there)

i pushed a change that i forgot to include (skips RNG initialization stuff if dropout isn't being used), there's still a small-ish gap on T4 though - i'll look into this.

> (2) The forward part has been [upstreamed to CUTLASS](https://github.com/NVIDIA/cutlass/tree/master/examples/41_fused_multi_head_attention), and @hwu36 is working on optimising it. It would be great to avoid too much divergence between both files, we need to figure out something.

luckily the changes in forward are much less involved than in backward, but willing to make changes to make the eventual merge easier.

> (3) Ease of use (eg compile time + binary size): as this multiplies by 2 the total number of attention kernels with the additional Dropout specialisation

More of a mitigation than a solution but one thing i noticed is that [nvcc compilation only uses 4 threads](https://github.com/facebookresearch/xformers/blob/main/setup.py#L248). when i was working on this i changed it to `os.environ.get("NVCC_THREADS", "4") and used this script to recompile

```
touch $(ls xformers/components/attention/csrc/cuda/mem_eff_attention/kernels/*.cu)
touch $(ls xformers/components/attention/csrc/cuda/mem_eff_attention/*.cu | grep generic)

XFORMERS_DISABLE_FLASH_ATTN=1 NVCC_THREADS=0 python3 setup.py develop
```

0 threads means nvcc uses as many as it wants, and that got things down to ~3min to recompile. can maybe also add some flags to control whether or not the dropout kernels get compiled, for people that aren't interested in working on those

> (4) Compatibility with older hardware (eg not just Sm75+)

i just realized i forgot to implement iterateCols for volta, let me go fix that. i'll see if its possible to trick the dispatcher to run kernels for older architectures so i can test the code for those as well

> EDIT: We seem to have some build errors on Windows that would need to be addressed (see CI)

will take a look

Reply:
Don't worry too much about Volta for now, but please make sure it builds at least. You should be able to build for Volta on any device via `TORCH_CUDA_ARCH_LIST`, so your code should build with `TORCH_CUDA_ARCH_LIST="6.0 6.1 7.0 7.5 8.0 8.6"`

Reply:
whoops meant to put the response above somewhere else

> Might need an additional guard for older GPUs (P100) where GEMMs don't already have this alignment requirement

i added some code in the `supports(...)` python method to prevent this kernel from being used if there's attention bias and its not aligned to allow 128 bit loads, is something like that what you are talking about here?

Reply:
Oh I didn't see that - yes it's perfect